backend/src/services/common/MeanTimeToRecovery.ts

```diff
@@ -0,0 +1,50 @@
+import { DeployInfo, DeployTimes } from "../../models/pipeline/DeployTimes";
+import {
+  AvgMeanTimeToRecovery,
+  MeanTimeToRecovery,
+  MeanTimeToRecoveryOfPipeline,
+} from "../../contract/GenerateReporter/GenerateReporterResponse";
+
+export function calculateMeanTimeToRecovery(
+  deployTimes: DeployTimes[]
+): MeanTimeToRecovery {
+  const meanTimeRecoveryPipelines: MeanTimeToRecoveryOfPipeline[] = deployTimes.map(
+    (item) => {
+      const totalTimeToRecovery = item.failed.reduce(
+        (prev: number, failedJob: DeployInfo): number => {
+          const failedJobCreateTime = new Date(
+            failedJob.pipelineCreateTime
+          ).getTime();
+          let timeToRecovery = Infinity;
+          for (let i = 0; i < item.passed.length; i++) {
+            const time =
+              new Date(item.passed[i].pipelineCreateTime).getTime() -
+              failedJobCreateTime;
+            timeToRecovery =
+              time > 0 && time < timeToRecovery ? time : timeToRecovery;
+          }
+          return prev + timeToRecovery;
+        },
+        0
+      );
+      return new MeanTimeToRecoveryOfPipeline(
+        item.pipelineName,
+        item.pipelineStep,
+        totalTimeToRecovery / item.failed.length || 0
```

Comment:
I just found a case are missed here.

Given 3 deployments of a pipeline: failed 1, failed 2, passed 3

MTTR should be: `passed 3 - failed 1`

According to the implementation: it will be `((passed 2 - failed 1) + (passed 3 - failed 1)) / 2`
Should add more test cases to cover this case.

Reply:
I'm not sure the accurate definition of recovery time, but I think `recovery` means `jobFinishTime` of passed deployment rather than its `pipelineCreateTime`.

Reply:
I also suggest sorting all failed and passed deployments by `pipelineCreateTime`, then to find every leading failed deployment and next passed deployment. 
For example, given sorted deployments: `failed 1`, `failed 2`, `passed 3`, `passed 4`, `failed 5`, `failed 6`, `passed 7`.
There are 2 recoveries: `failed 1` and `passed 3`, `failed 5` and `passed 7`.

Reply:
Hi @samuraime I've changed the algorithm for MTTR as you suggested which is correct.  But the time use for calculate I think `jobFinishTime` or `pipelineCreateTime` won't make much difference because the pipeline build  doesn't take long time(most less than 10min) compare to recovery time (usually need hours). Feel free to contact me through wechat if you have different opinions.